tests/python/pants_test/engine/test_isolated_process.py

```diff
@@ -84,22 +154,125 @@ def java_sources_to_javac_args(sources_snapshot, out_dir):
                                    directories_to_create=(out_dir.path,))
 
 
-def process_request_from_java_sources(binary):
-  env = []
-  return ExecuteProcessRequest(
-    argv=binary.prefix_of_command() + ('-version',),
-    env=env,
-    input_files_digest="e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
+class JavacVersionCommand(Javac):
+
+  def gen_argv(self, snapshots=None):
+    if snapshots:
+      raise ValueError("JavacVersionCommand cannot use input snapshots '{}'"
+                       .format(snapshots))
+    return (self.bin_path, '-version',)
+
+
+EMPTY_FINGERPRINT = 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855'
+
+
+@rule(ExecuteProcessRequest, [Select(JavacVersionCommand)])
+def process_request_from_javac_version(javac_version_command):
+  yield ExecuteProcessRequest(
+    argv=javac_version_command.gen_argv(),
+    env=[],
+    input_files_digest=EMPTY_FINGERPRINT,
     digest_length=0)
 
 
-def process_request_java_args_from_java_sources(binary, sources_snapshot, out_dir):
-  env = []
-  return ExecuteProcessRequest(
-    argv=binary.prefix_of_command() + tuple(f.path for f in sources_snapshot.files),
-    input_files_digest=str(sources_snapshot.fingerprint),
-    digest_length=sources_snapshot.digest_length,
-    env=env)
+class JavacVersionOutput(datatype('JavacVersionOutput', [
+    'exit_code',
```

Comment:
exit_code seems like a strange thing to be modelling here as a field; wouldn't this be more naturally exposed by making the task Throw on a non-zero exit code?

I can imagine tasks where we want to expose exit_code, but I don't think invoking javac is one of them...

(Can happily be a follow-up to experiment separately with failure-related things)

Reply:
I hadn't thought of it this way, I think that change would be reasonable to introduce in this PR.

Reply:
Modified JavacVersionOutput to only contain the `stderr` field and instead raise if the exit code is nonzero, and caught the Throw in the test!